src/openvic-simulation/scripts/Condition.cpp

```diff
@@ -0,0 +1,659 @@
+	ret &= add_condition("has_pop_type", IDENTIFIER, PROVINCE, UNSPECIFIED, NONE, POP_TYPE);
+	ret &= add_condition("has_province_flag", IDENTIFIER, PROVINCE, UNSPECIFIED, NONE, PROVINCE_FLAG);
+	ret &= add_condition("has_province_modifier", IDENTIFIER, PROVINCE, UNSPECIFIED, NONE, MODIFIER_EFFECT);
+	ret &= add_condition("has_recent_imigration", INTEGER, PROVINCE); //paradox typo
+	ret &= add_condition("is_blockaded", BOOLEAN, PROVINCE);
+	ret &= add_condition("is_capital", BOOLEAN, PROVINCE);
+	ret &= add_condition("is_coastal", BOOLEAN, PROVINCE);
+	ret &= add_condition("is_colonial", BOOLEAN, PROVINCE);
+	ret &= add_condition("is_overseas", BOOLEAN, PROVINCE);
+	ret &= add_condition("is_state_capital", BOOLEAN, PROVINCE);
+	ret &= add_condition("life_rating", REAL, PROVINCE);
+	ret &= add_condition("minorities", BOOLEAN, PROVINCE);
+	ret &= add_condition("owned_by", IDENTIFIER, PROVINCE, UNSPECIFIED, NONE, COUNTRY_TAG);
+	ret &= add_condition("port", BOOLEAN, PROVINCE);
+	ret &= add_condition("province_control_days", INTEGER, PROVINCE);
+	ret &= add_condition("province_id", IDENTIFIER, PROVINCE, UNSPECIFIED, NONE, PROVINCE_ID);
+	ret &= add_condition("region", IDENTIFIER, PROVINCE, UNSPECIFIED, NONE, REGION);
+	ret &= add_condition("state_id", IDENTIFIER, PROVINCE, UNSPECIFIED, NONE, PROVINCE_ID);
+	ret &= add_condition("terrain", IDENTIFIER, PROVINCE, UNSPECIFIED, NONE, TERRAIN);
+	ret &= add_condition("trade_goods", IDENTIFIER, PROVINCE, UNSPECIFIED, NONE, TRADE_GOOD);
+	ret &= add_condition("unemployment_by_type", GROUP, PROVINCE);
+	ret &= add_condition("units_in_province", INTEGER, PROVINCE);
+
+	/* Pop Scope Conditions */
+	ret &= add_condition("agree_with_ruling_party", REAL, POP);
+	ret &= add_condition("cash_reserves", REAL, POP);
+	ret &= add_condition("consciousness", REAL, POP);
+	ret &= add_condition("culture", IDENTIFIER, POP, UNSPECIFIED, NONE, CULTURE);
+	ret &= add_condition("everyday_needs", REAL, POP);
+	ret &= add_condition("has_pop_culture", IDENTIFIER, POP, UNSPECIFIED, NONE, CULTURE);
+	ret &= add_condition("has_pop_religion", IDENTIFIER, POP, UNSPECIFIED, NONE, RELIGION);
+	ret &= add_condition("life_needs", REAL, POP);
+	ret &= add_condition("luxury_needs", REAL, POP);
+	ret &= add_condition("militancy", REAL, POP);
+	ret &= add_condition("political_movement", REAL, POP);
+	ret &= add_condition("pop_majority_issue", IDENTIFIER, POP, UNSPECIFIED, NONE, ISSUE);
+	ret &= add_condition("religion", IDENTIFIER, POP, UNSPECIFIED, NONE, RELIGION);
+	ret &= add_condition("social_movement", BOOLEAN, POP);
+	ret &= add_condition("strata", IDENTIFIER, POP, UNSPECIFIED, NONE, POP_STRATA);
+	ret &= add_condition("type", IDENTIFIER, POP, UNSPECIFIED, NONE, POP_TYPE);
+
+	#define IMPORT_IDENTIFIERS(fun, value_type, scope, scope_change, key_id_type, val_id_type) \
+		for (std::string_view id : fun) { ret &= add_condition(id, value_type, scope, scope_change, key_id_type, val_id_type); }
```

Comment:
```suggestion
	const auto import_identifiers = [this, &ret](
		std::vector<std::string_view> const& identifiers, value_type_t value_type, scope_t scope,
		scope_t scope_change = scope_t::UNSPECIFIED,
		identifier_type_t key_identifier_type = identifier_type_t::NONE,
		identifier_type_t value_identifier_type = identifier_type_t::NONE) -> void {
		for (std::string_view identifier : identifiers) {
			ret &= add_condition(
				identifier, value_type, scope, scope_change, key_identifier_type, value_identifier_type
			);
		}
	};
```
Spartan's been advocating against using macros where lambdas also work, and in general they are better, mainly in terms of having specific code that an error message can refer to. In this case they also allow default arguments, e.g. all but one use of `IMPORT_IDENTIFIERS` have their last argument as `NONE`.

Reply:
Spartan is right -- I've just been using macros because they are faster to write and for a prototype speed > correcteness.